src/hardware/iohandler.cpp

```diff
@@ -27,52 +27,93 @@
 
 //#define ENABLE_PORTLOG
 
-IO_WriteHandler * io_writehandlers[3][IO_MAX];
-IO_ReadHandler * io_readhandlers[3][IO_MAX];
+std::unordered_map<io_port_t, IO_WriteHandler> io_writehandlers[IO_SIZES] = {};
+std::unordered_map<io_port_t, IO_ReadHandler> io_readhandlers[IO_SIZES] = {};
 
-static Bitu IO_ReadBlocked(Bitu /*port*/,Bitu /*iolen*/) {
-	return ~0;
+void port_within_proposed(io_port_t port) {
+	assert(port < std::numeric_limits<io_port_t_proposed>::max());
```

Comment:
This is a really interesting point.

I concluded that it has to be `<` given that `port` is the lookup key into the `map`, so with the proposed `uint16_t` we'd have a key range from 0 to 65535 inclusive.  Where as `<=` would allow port (and thus key) to be 65536 which would roll around to 0 giving us two keys - 0 and 65536 that point to the same IO handler -- which I don't think we want. 

Dosbox also constrains the port range to 0 to 65535 when the port value is fed into the CPU core in these functions:

 - IO_WriteB() { ... reg_dx = static_cast<uint16_t>(port); }
 - IO_WriteW() { ... reg_dx = static_cast<uint16_t>(port); }
 - IO_WriteD() { ... reg_dx = static_cast<uint16_t>(port); }
 - IO_ReadB() { ... reg_dx = static_cast<uint16_t>(port); }
 - IO_ReadW() { ... reg_dx = static_cast<uint16_t>(port); }
 - IO_ReadD() { ... reg_dx = static_cast<uint16_t>(port); }

When I saw this code, this is how I knew that `port` should just be a `uint16_t` and not a BitU.  If DOS really tried to set ports above >= the size of a uint16_t then this code would roll them around and mangle the results. 

But thinking about your question from the DOS implementation side: do ports range from 1 to 65536, and are not actually zero-based?  Is it legal to have a "port 0"? I'm not sure!

Reply:
I'm having a hard time finding the actual API specifically for IO port values, however I did find an an exhaustive list of actual ports and their assignments:

http://bochs.sourceforge.net/techspec/PORTS.LST

They indeed are zero based and grew to `E2E8-E2EF`, which was assigned to "8514/A and compatible video cards (e.g. ATI Graphics Ultra)".

It's fascinating to see the port addresses being consumed by newer and newer hardware in the list. This tells me the DOS era died out just in time before they ran out of port numbers :-)